src/perf_counters.cc

```diff
@@ -71,7 +71,7 @@ bool PerfCounters::IsCounterSupported(const std::string& name) {
   return (ret == PFM_SUCCESS);
 }
 
-PerfCounters PerfCounters::Create(
+std::shared_ptr<PerfCounters> PerfCounters::Create(
```

Comment:
i would expect (maybe naively) that we only create the perf counters once.  why, then, would we need a shared pointer here?  either the created perf counter object is owned by this object _or_ the caller (which should be singular).

Reply:
I understand that. However what we had before with a reference counter was effectively a poor man's intrusive shared pointer like boost::intrusive_ptr. That was the reason I used std::shared_ptr. However that might have been only because PerfCounters was static and shared across the board with several threads. If we use std::unique_ptr then what we are saying is that this base object is never supposed to be shared, which is a huge assumption I did not want to make at this point. But I can put it there and test.

Reply:
The intention is not to share counters, in the absence of a scenario. Unless we have a scenario for google/benchmarks, I'd suggest we maintain the invariant.

Reply:
The way it was, it was written to share counters, even with a specific reference counter in there. As a suggestion, I changed PerfCounters to be a regular member. Have a look at the commit just pushed and see if you like. It simplifies things quite a bit.

Reply:
Sorry, I misspoke. You are right that the `PerfCounters` object is shared. What I meant to say is that we want one instance only, to be shared by all Measurements. IIUC, this PR would create a new `PerfCounters` for each `PerfCountersMeasurement`.

We could get rid of the ref count like you suggesting by keeping `counters_` static, and making sure it's inited only once. But, one drawback of using a pointer there is that now measurements would have to be through an indirection. That could add noise to measurements - dcache misses, for instance - and, given the purpose of this code, I'd argue it's best to avoid such noise as much as possible.

Reply:
Ohhh ok now I am getting the reason why these objects were made static. It makes sense. Although I would like to propose this approach: why not create a single PerfCountersMeasurement object and pass it to each BenchmarkRunner's constructor? That would be a bit more explicit than a static object.

Reply:
a `PerfCountersMeasurement` keeps track of the start/stop counter values - "start" and "stop" being values snapshotted at the beginning/end of some region of code of interest (==benchmarked code). Arguably, we'd want separate instances of `PerfCountersMeasurement` for separate experiments - they can share, though, the same source of info (i.e. the `PerfCounters` object)

Reply:
I cannot think of a use case for having overlapping measurements of the same counters. If we have multiple threads running separate tests on the same process, that is outright wrong because the samples will contain contributions from all threads. If we have several tests running sequentially then having separate values for each one might affect adversely (say dTLB) than having a single, shared PerfCountersMeasurement object.

Reply:
Right, I'm not suggesting either of those. It may be more of an aesthetic consideration - you want to make a measurement, you make a `PCM` object... but I can see how what you're suggesting could let us assert that, for example, only one client makes measurements, and I like that. Maybe it'd be easier to see how the code looked like?

Reply:
Yes, Sir! I have added another commit to this patch. Please have a look. I tried to describe as much as possible in the commit but please let me know your thoughts. All tests are passing but I am afraid of some side effects I'm not taking in account for my lack of deep understanding of the project.